vncviewer/vncviewer.cxx

```diff
@@ -144,18 +144,36 @@ void about_vncviewer()
   fl_message("%s", about_text());
 }
 
-void run_mainloop()
+static int run_mainloop(const char* vncserver, network::Socket* sock)
 {
-  int next_timer;
+  if(exitError != NULL) {
+    free(exitError);
+    exitError = NULL;
+  }
 
-  next_timer = Timer::checkTimeouts();
-  if (next_timer == 0)
-    next_timer = INT_MAX;
+  CConn *cc = new CConn(vncServerName, sock);
+  int next_timer;
 
-  if (Fl::wait((double)next_timer / 1000.0) < 0.0) {
-    vlog.error(_("Internal FLTK error. Exiting."));
-    exit(-1);
+  inMainloop = true;
+  exitMainloop = false;
+  while (!exitMainloop) {
+    next_timer = Timer::checkTimeouts();
+    if (next_timer == 0)
+      next_timer = INT_MAX;
+
+    if (Fl::wait((double)next_timer / 1000.0) < 0.0) {
+      vlog.error(_("Internal FLTK error. Exiting."));
+      exit(-1);
+    }
   }
+  inMainloop = false;
+
+  delete cc;
+
+  if (exitError != NULL && alertOnFatalError)
```

Comment:
I'm not sure this is a safe principle. We've not previously made the distinction between errors that are fatal for the connection, and errors that are fatal for the entire process.

Have you gone through the calls to `exit_vncviewer()` and see how they are used?

We should probably also split `exit_vncviewer()` to two functions to handle different levels of errors.